Refactor metadata fetching into `MetadataProvider` trait

csssuf opened the pull request:
Instead of `fetch_metadata()` fetching all potentially needed metadata, return a `MetadataProvider` trait object which has drop-in replacements for `Metadata`'s functions. This allows only the requested metadata types to be fetched and avoids potentially expensive fetching work.

There's still a little bit of cleanup work that could take place in some of the providers (mostly rolling now-duplicate functions into the trait implementation), but those seemed like they should wait for another PR. This PR contains roughly the minimal amount of refactoring necessary to switch over to the `MetadataProvider` trait.

Fixes coreos/bugs#2362.

csssuf committed:
providers: add MetadataProvider trait

Add a new trait `MetadataProvider` with functions for fetching each type
of metadata coreos-metadata knows about, and helper functions for
writing metadata to disk.

csssuf committed:
providers/*: implement MetadataProvider trait

Add an implementation of `MetadataProvider` for each existing metadata
provider.

csssuf committed:
lib: use MetadataProvider trait for fetch_metadata

Instead of fetching all possible metadata in fetch_metadata(), return a
trait object which provides the same functions for writing to disk as
the original `Metadata` struct, but only fetches metadata when
necessary.

csssuf committed:
providers/*: drop fetch_metadata() functions

Drop the unused fetch_metadata() function from each provider, as the
`MetadataProvider` trait is now used in their place.

csssuf committed:
metadata: drop metadata module

`Metadata` is now unused after the switch to `MetadataProvider`, so drop
it.

crawford commented on src/lib.rs:
```diff
@@ -85,20 +85,24 @@ use metadata::Metadata;
 /// function to fetch metadata. The configured provider is passed in and this
 /// function dispatches the call to the correct provider-specific fetch function
-pub fn fetch_metadata(provider: &str) -> Result<Metadata> {
+pub fn fetch_metadata(provider: &str) -> Result<Box<providers::MetadataProvider>> {
```
Why is the Box needed?

csssuf replied:
Since we're returning multiple different concrete types that all implement `MetadataProvider`, `impl Trait` can't be used and we need a `Box` to create an actual trait object.

crawford replied:
I thought that was the whole point of `impl Trait`...

csssuf replied:
Unfortunately not. From [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md#semantics):
>If a function returns impl Trait, its body can return values of any type that implements Trait, but all return values need to be of the same type.

lucab replied:
@csssuf for my own reference: pushing the `-> impl Trait` one more level down too (i.e. returned by `new()`), would that also not work?

csssuf replied:
[Nope](https://play.rust-lang.org/?gist=5adf057156d6a29d6ab2791d314a546b&version=stable&mode=debug). The compiler needs to be able to figure out static dispatch to use `impl Trait`, so there's no way to trick it into doing the thing we want here.

cgwalters replied:
https://www.reddit.com/r/rust/comments/8sqiin/newbie_question_why_is_the_impl_and_dyn_in_the/ is also useful on this topic (there's various others too)

With rustc beta today for example, one can write:

```
diff --git a/src/lib.rs b/src/lib.rs
index 0f80d1c..2ec3af3 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -90,7 +90,7 @@ macro_rules! box_result {
 /// `fetch_metadata` is the generic, top-level function that is used by the main
 /// function to fetch metadata. The configured provider is passed in and this
 /// function dispatches the call to the correct provider-specific fetch function
-pub fn fetch_metadata(provider: &str) -> Result<Box<providers::MetadataProvider>> {
+pub fn fetch_metadata(provider: &str) -> Result<Box<dyn providers::MetadataProvider>> {
     match provider {
         "azure" => box_result!(azure::Azure::new()),
         "cloudstack-metadata" => box_result!(cloudstack::network::CloudstackNetwork::new()),
```

Then it's explicitly clear the goal is dynamic dispatch.

BTW I love how much discussion there's been over a single heap allocation :smile:

sdemos replied:
without having looked into it at all, I wonder if it would be possible to simultaneously use `dyn` and `impl Trait` to do what we want without a heap allocation? mostly curious, the current version is totally fine by me too.

csssuf replied:
No, they're mutually exclusive. We _could_ avoid the `Box` by using an enum as the function's return type, and essentially provide our own dynamic dispatch by implementing the trait for the enum, but IMO that's even less clean than simply returning a `Box`ed trait object.

csssuf committed:
providers/ec2: use Ec2Provider in mock tests

Now that we expose an Ec2Provider struct in providers/ec2, use it in
the ec2 mock tests.

lucab commented on src/providers/azure/mod.rs:
```diff
@@ -275,19 +279,36 @@ impl Azure {
+impl MetadataProvider for Azure {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let attributes = self.get_attributes()?;
+        let mut out = HashMap::new();
```
Here and in all the other providers too, what about `with_capacity(2)` instead?

lucab commented on src/providers/cloudstack/configdrive.rs:
```diff
@@ -69,36 +82,67 @@ impl ConfigDrive {
+impl ::std::ops::Drop for ConfigDrive {
+    fn drop(&mut self) {
+        if self.temp_dir.is_some() {
+            ConfigDrive::unmount(&self.path).unwrap();
```
I'm not sure we should panic here. (This is carried-over code, feel free to defer this)

sdemos commented on src/providers/gce/mod.rs:
```diff
@@ -14,82 +14,122 @@
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::with_capacity(3);
+
+        let add_value = |map: &mut HashMap<_, _>, key: &str, name| -> Result<()> {
```
Unfortunate that there is so much duplication with the `add_value` closure, but I'm not sure if there is a better way to do it after thinking through some of the alternatives.

sdemos reviewed (approved):
lgtm after we get positive tests on all the major cloud providers (it might need a rebase on master, shouldn't be any conflicts though)
